tests/integration/api_exec_test.py

```
@@ -75,6 +75,75 @@ def test_exec_command_streaming(self):
+            'echo hello err >&2'])
+        cmd = 'sh -c "{}"'.format(script)
+
+        # tty=False, stream=False, demux=False
+        res = self.client.exec_create(id, cmd)
+        exec_log = self.client.exec_start(res)
+        assert exec_log == b'hello out\nhello err\n'
+
+        # tty=False, stream=True, demux=False
+        res = self.client.exec_create(id, cmd)
+        exec_log = self.client.exec_start(res, stream=True)
+        assert next(exec_log) == b'hello out\n'
+        assert next(exec_log) == b'hello err\n'
+        with self.assertRaises(StopIteration):
+            next(exec_log)
+
+        # tty=False, stream=False, demux=True
+        res = self.client.exec_create(id, cmd)
+        exec_log = self.client.exec_start(res, demux=True)
+        assert exec_log == (b'hello out\n', b'hello err\n')
+
+        # tty=False, stream=True, demux=True
+        res = self.client.exec_create(id, cmd)
+        exec_log = self.client.exec_start(res, demux=True, stream=True)
+        assert next(exec_log) == (b'hello out\n', None)
+        assert next(exec_log) == (None, b'hello err\n')
+        with self.assertRaises(StopIteration):
+            next(exec_log)
+
+        # tty=True, stream=False, demux=False
+        res = self.client.exec_create(id, cmd, tty=True)
+        exec_log = self.client.exec_start(res)
+        assert exec_log == b'hello out\r\nhello err\r\n'
+
+        # tty=True, stream=True, demux=False
+        res = self.client.exec_create(id, cmd, tty=True)
+        exec_log = self.client.exec_start(res, stream=True)
+        assert next(exec_log) == b'hello out\r\n'
+        assert next(exec_log) == b'hello err\r\n'
+        with self.assertRaises(StopIteration):
+            next(exec_log)
+
+        # tty=True, stream=False, demux=True
+        res = self.client.exec_create(id, cmd, tty=True)
+        exec_log = self.client.exec_start(res, demux=True)
+        assert exec_log == (b'hello out\r\nhello err\r\n', b'')
```

This is a bit unfortunate. To be consistent, this should probably be:

```python
assert exec_log == (b'hello out\r\nhello err\r\n', None)
```

But on the other hand, `tty=True` and `demux=True` does not really make sense _anyway_ so I doubt it's going to be an issue in practice.

---

Should we raise an exception if someone tries to use `demux` and `tty` together?
Like https://github.com/docker/docker-py/blob/master/docker/api/container.py#L1101-L1104 but in reverse.

---

We could but strictly speaking I don't think it's really _invalid_. I don't see a use case for it, but I'd like not to restrict the API. I propose we fix the consistency issue instead. Here is a possible fix: https://github.com/docker/docker-py/pull/2150/commits/f80eecc0a3dc0be580b1a48259938ca98de1598d